### PR TITLE
Tighten TextIOWrapper buffer types

### PR DESCRIPTION
**Pull request by srittau**
This matches the documentation, which says:

> A buffered text stream providing higher-level access to a BufferedIOBase
> buffered binary stream.

Cf #10229

### CODE REVIEW

**Comment by srittau** [comment]
Also, the current annotations are unsound. You can pass in any `IO[bytes]`, but are guaranteed to get back a `BinaryIO` from `.buffer` and `.detach()`, although in reality you just get back what you passed in.

**Comment by AlexWaygood** [comment]
I'd be interested to see the mypy_primer results if we tried making it generic. It might be very disruptive, but it also might not.

**Comment by srittau** [comment]
Anyway, the merge order is #10233 -> #10232 -> potentially this or #10229

**Comment by srittau** [comment]
I suggest we merge the two other PRs first, rerun the tests here and then I'll also do a test with a generic `TextIOWrapper`.

**Comment by srittau** [comment]
I merged the additional changes into #10229, closing this here.
